linkifiers: Add joint validation for linkifier urls and patterns.

akshatdalton opened the pull request:
This commit validates the linkifier urls and patterns together, and add
the following additional checks:

1. All groups in the pattern must be used in the url.
2. All groups in the url must be declared in the pattern.
3. Substituting all groups in the url must result in a valid url.

Fixes #16482.

We also allow substitutions in domain names now.


**Testing plan:** <!-- How have you tested? -->
New test cases are added to match the change.

andersk commented on zerver/lib/test_classes.py:
```diff
@@ -781,6 +781,34 @@ def assert_json_error(self, result: HttpResponse, msg: str, status_code: int=400
+
+        try:
+            json = orjson.loads(result.content)
+        except Exception:  # nocoverage
```
Be specific: `except orjson.JSONDecodeError`. But given that this is test code, it’s better not to handle this exception at all—if we get invalid JSON here, something has gone wrong and we *want* a stack trace.

akshatdalton replied:
If we want a stack trace then should I just write
`json_content = orjson.loads(result.content)` ?

andersk replied:
Yes.

andersk commented on zerver/lib/test_classes.py:
```diff
@@ -781,6 +781,34 @@ def assert_json_error(self, result: HttpResponse, msg: str, status_code: int=400
+            json = orjson.loads(result.content)
+        except Exception:  # nocoverage
+            json = {'errors': {}}
+        errors = json.get('errors')
```
This can be `None`. Since you’re not expecting that, use `json['errors']` rather than `json.get('errors')`.

akshatdalton replied:
Okay this should be simply `errors = json_content['errors']` since we want a stack trace?

andersk replied:
Yes.

andersk commented on zerver/lib/test_classes.py:
```diff
@@ -781,6 +781,34 @@ def assert_json_error(self, result: HttpResponse, msg: str, status_code: int=400
+        """
+
+        try:
+            json = orjson.loads(result.content)
```
`json` is the wrong variable name for something we’ve just translated *out* of JSON.

andersk commented on zerver/lib/test_classes.py:
```diff
@@ -781,6 +781,34 @@ def assert_json_error(self, result: HttpResponse, msg: str, status_code: int=400
+        errors = json.get('errors')
+        error_list: List[str] = []
+        for error_type in errors:
+            error_list = error_list + errors.get(error_type)
```
This loop takes quadratic time. Among the many ways to improve it, the simplest is replacing it with `sum(errors.values(), [])`.

andersk commented on zerver/lib/test_classes.py:
```diff
@@ -781,6 +781,34 @@ def assert_json_error(self, result: HttpResponse, msg: str, status_code: int=400
+            error_list = error_list + errors.get(error_type)
+        return error_list
+
+    def assert_all_json_errors(self, result: HttpResponse, expected_errors: List[str]) -> None:
```
If you’re not mutating `expected_errors`, use the immutable `Sequence` type rather than `List`.

andersk commented on zerver/models.py:
```diff
@@ -750,38 +750,63 @@ def flush_realm_emoji(sender: Any, **kwargs: Any) -> None:
+            )
+
+        group_set = set(pattern.groupindex.keys())
+        group_match_regex = re.compile(r'%\((?P<group_name>[^)]+)\)s')
```
This is not correct in the presence of `%%` (which is how you embed a literal `%` in a format string).

Group names aren’t allowed to include `(` either.

There’s no point in using `re.compile` on a regex you’re throwing away after one use.

akshatdalton replied:
> This is not correct in the presence of `%%` (which is how you embed a literal `%` in a format string).

Can you give an example so that I can understand this more better ?

andersk replied:
```pycon
>>> "Hello %(foo)s" % {"foo": "bar"}
'Hello bar'
>>> "Hello %%(foo)s" % {"foo": "bar"}
'Hello %(foo)s'
>>> "Hello %%%(foo)s" % {"foo": "bar"}
'Hello %bar'
>>> "Hello %%%%(foo)s" % {"foo": "bar"}
'Hello %%(foo)s'
>>> "Hello %%%%%(foo)s" % {"foo": "bar"}
'Hello %%bar'
>>> "Hello %(inv(alid)s" % {"inv(alid": "bar"}
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: incomplete format key
```

andersk commented on zerver/models.py:
```diff
@@ -750,38 +750,63 @@ def flush_realm_emoji(sender: Any, **kwargs: Any) -> None:
+            group_set.remove(group_name)
+
+        # Report patterns missing in URL format string.
+        if len(group_set):
```
PEP 8 style is just `if group_set`.

andersk commented on zerver/models.py:
```diff
@@ -750,38 +750,63 @@ def flush_realm_emoji(sender: Any, **kwargs: Any) -> None:
+        if len(group_set):
+            # We just report the first missing pattern here. People can incrementally
+            # resolve errors if there are multiple missing patterns.
+            group_list = list(group_set)
+            group_list.sort()
```
`group_list = sorted(group_set)` is simpler.

andersk commented on zerver/models.py:
```diff
@@ -750,38 +750,63 @@ def flush_realm_emoji(sender: Any, **kwargs: Any) -> None:
+            group_list.sort()
+            group_name = group_list[0]
+            raise ValidationError(
+                _("Group '%(name)s' in filter pattern is not present in URL format string."),
```
Use `"… %(name)r …"` instead of `"… '%(name)s' …"` for correct escaping.

akshatdalton commented:
@andersk can you please answer to the comments, I have added ?

akshatdalton commented:
I have only pushed the commit: `linkifiers: Add joint validation for linkifier urls and patterns` since issue #16482 is labelled as: **`priority: high`**.
I will fix the issues mentioned in: https://github.com/zulip/zulip/issues/16482#issuecomment-710522459, #12323 and #12239 in separate commits. This commit aims to fix issue #16482.

Following are the points I have addressed in my current push:

- This commit adds the joint validation for linkifier urls and patterns.
- Removed the field level validator for `pattern` and its validation is done inside the `clean` method, (Related [comment](https://github.com/zulip/zulip/pull/13398/files#r506765559)). As a result of this, `pattern` validation error will be collected in [`__all__`](https://github.com/zulip/zulip/blob/367cf897fa32262a41cbcd6527a3b319e73ec776/static/js/settings_linkifiers.js#L56-L58) instead of [`pattern`](https://github.com/zulip/zulip/blob/367cf897fa32262a41cbcd6527a3b319e73ec776/static/js/settings_linkifiers.js#L48-L50). Using [`pattern`](https://github.com/zulip/zulip/blob/367cf897fa32262a41cbcd6527a3b319e73ec776/static/js/settings_linkifiers.js#L48-L50) to collect field-level validation error still counts, since there can be cases like `Pattern` field is left blank or input to such field is not a string.
- Used `set` instead of `list` to improve the optimization (Related [comment](https://github.com/zulip/zulip/pull/13398/commits/45cc26e49b90b046ce2ba14378428a167e803c60#r506772248)).
Now, since `set` is used, I'm not using `set.remove()` method instead I'm [constructing a new set](https://github.com/zulip/zulip/pull/16860/files#diff-c210d4a13c8c2d0d481c3755996fb1dfe0dadd76cc1bcba6e817ee9bfa29c7e0R950) and taking the [set difference](https://github.com/zulip/zulip/pull/16860/files#diff-c210d4a13c8c2d0d481c3755996fb1dfe0dadd76cc1bcba6e817ee9bfa29c7e0R962) between [`group_set`](https://github.com/zulip/zulip/pull/16860/files#diff-c210d4a13c8c2d0d481c3755996fb1dfe0dadd76cc1bcba6e817ee9bfa29c7e0R949) and [`found_group_set`](https://github.com/zulip/zulip/pull/16860/files#diff-c210d4a13c8c2d0d481c3755996fb1dfe0dadd76cc1bcba6e817ee9bfa29c7e0R950).
The reason behind this is there can be a case like when the `Pattern`: `#(?P<id>[0-9]+)` and the `url_string_format`: `https://realm.com/my_realm_filter/?value1=%(id)s&value2=%(id)s`, in such a case `clean` method will throw an error.
- [`URLValidation` for substituted string](https://github.com/zulip/zulip/issues/12323#issue-443591221) will be done when we will remove field-level validator: `filter_format_validator` to meet the conditions mentioned [here](https://github.com/zulip/zulip/issues/12239#issue-439049291) (will be done in separate commit since this commit aims to fix the issue #16482).


I don't understand:
> This is not correct in the presence of `%%` (which is how you embed a literal `%` in a format string).

how the [above comment](https://github.com/zulip/zulip/pull/16860#discussion_r539112625) is valid here?
The [examples provided](https://github.com/zulip/zulip/pull/16860#discussion_r581554647) should be valid for https://github.com/zulip/zulip/pull/13398/files#diff-fdf42dee6384fb054fbc027e0b7b4f68896624321adb1b5f7a4ccf89094b92edR1630-R1633 and not for this commit.

And even in such case, how will it [break the rendering](https://github.com/zulip/zulip/pull/13398/files#r506773816)? [Such function](https://github.com/zulip/zulip/pull/13398/files#diff-fdf42dee6384fb054fbc027e0b7b4f68896624321adb1b5f7a4ccf89094b92edR1631) interpolates to use [`string.format()`](https://realpython.com/python-string-formatting/#2-new-style-string-formatting-strformat) method and not
[`% formatting method`](https://realpython.com/python-string-formatting/#1-old-style-string-formatting-operator).

@timabbott @andersk, please review.

akshatdalton commented:
(Just resolved some conflicts and added the attribute to the original author.)

@timabbott, a gentle reminder about this PR.

akshatdalton commented:
The puppeteer test is still failing at `test_edit_invalid_linkifier`. 😕
@Riken-Shah, FYI.

akshatdalton commented:
(Just resolved some conflicts and added attribute to the original author.)
Major changes mentioned in: https://github.com/zulip/zulip/pull/16860#issuecomment-823295327.

@timabbott, a gentle reminder about this PR.

timabbott commented:
OK, I merged this after the following changes:
* A bit of cleanup of the `clean()` function to be more readable in terms of structure and comments.
* Added a few new tests, including one for the `%%` case that was discussed above, where our behavior is to incorrectly reject a URL that should be valid.  We'll want to fix this eventually, but it's not important and is I think in scope for the existing #12323 issue.

Thanks @akshatdalton!  Even though this doesn't complete our ambitions here, it will definitely avoid some amount of unpleasant bugs that users have been running into.  And setting this PR should unblock the rest of @aero31aero's work that had been stacked on top of this in his earlier PRs :).
